mautic/mautic
# Unsubscribe improvement

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y 
| Automated tests included? | Y 
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | /
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description 

⚠️ ~~This needs `FromEmailHelper` service that is introduced in https://github.com/mautic/mautic/pull/7811 so it must be merged first.~~

**Acceptance Criteria**

- User is able to unsubscribe from an old email, even if the specific email stat has been deleted
Contact identifier is hashed
- The unsubscribe URLs are generated using the hash stored in an email stat. If the email stat is deleted, the unsubscribe URL will no longer function as it will not be able to find the contact to unsubscribe.

We should change the format of the unsubscribe URL to also include a means to identify the contact to unsubscribe in the case the stat hash is not found. For example /email/unsubscribe/$STAT_HASH/$EMAIL_ADDRESS/$SECRET_HASH In Mautic's routing config.

$EMAIL_ADDRESS/$SECRET_HASH would be made optional for BC purposes so that Symfony continues to recognize the old URL. If a stat is not found, we then find the contact using $EMAIL_ADDRESS/$SECRET_HASH.

$SECRET_HASH is just for verification so anyone who would know just the email address cannot unsubscribe it.

The $SECRET_HASH would be generated from the instance's secret_key and the email address. Hash the email address using the secret_key. This hashing mechanism should be the best, but make a quick research: https://www.php.net/manual/en/function.hash-hmac.php. (sha256) It must be secure and fast.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Send email with unsubscribe link
3. Click link - it should look like /email/unsubscribe/$STAT_HASH/$EMAIL_ADDRESS/$SECRET_HASH
4. It should unsubscribe you even if:
 4.a. you use just /email/unsubscribe/$STAT_HASH
 4.b. you use wrong hash but right email with right $SECRET_HASH

#### Other areas of Mautic that may be affected by the change:
1. Email unsubscribe only


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10846"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
